### PR TITLE
feat: support conditional operators

### PR DESCRIPTION
**Pull request by AlexVarchuk**
## What/Why/How?
add support conditional operators for openApi 3.1. Add supporting `x-displayName` for if/else.

## Reference
```yaml
properties:
  photoUrls:
    description: The list of URL to a cute photos featuring pet
    type: [string, integer, 'null', array]
    xml:
      name: photoUrl
      wrapped: true
    items:
      type: string
      format: url
    if:
      x-displayName: isString
      type: string
    then:
      minItems: 1
      maxItems: 10
    else:
      x-displayName: notString
      type: [integer, 'null', array]
      minItems: 1
      maxItems: 20
```

```yaml
properties:
  userStatus:
    description: User status
    type: integer
    format: int32
if:
  title: userStatus === 10
  properties:
    userStatus:
      enum: [10]
then:
  required: ['phone']
else:
  required: []
```

## Screenshots (optional)
Screenshot 2022-05-16 at 18 56 00
Screenshot 2022-05-16 at 18 56 15
Screenshot 2022-05-16 at 18 56 29

## Check yourself

- [x] Code is linted
- [x] Tested
- [x] All new/updated code is covered with tests

### CODE REVIEW

File: src/services/models/Schema.ts [hunk 1]
```diff
@@ -152,6 +154,10 @@ export class SchemaModel {
       return;
     }
 
+    if (schema.if || schema.then || schema.else) {
```
Comment (zalesky) [inline src/services/models/Schema.ts:157]: I think we should union this with `if ((!ifOperator && !thenOperator) || (!ifOperator && !elseOperator)) return;` from `initConditionalOperators`
I mean better to have 1 condition in 1 place

Comment (AlexVarchuk) [reply src/services/models/Schema.ts]: @zalesky Thank for good suggestion. Improved.

**Comment by AlexVarchuk** [comment]
after merge [#1942]( https://github.com/Redocly/redoc/pull/1942) unit test should be passed

**Review by zalesky — APPROVED** [review]

File: src/services/OpenAPIParser.ts [hunk 2]
```diff
@@ -279,7 +279,19 @@ export class OpenAPIParser {
       }
 
       if (subSchema.type !== undefined) {
-        receiver.type = subSchema.type;
+        if (Array.isArray(subSchema.type) && Array.isArray(receiver.type)) {
+          receiver.type = subSchema.type.concat(...receiver.type);
```
Comment (RomanHotsiy) [inline src/services/OpenAPIParser.ts:283]:
```suggestion
          receiver.type = [...subSchema.type, ...receiver.type];
```

File: src/services/OpenAPIParser.ts [hunk 3]
```diff
@@ -279,7 +279,19 @@ export class OpenAPIParser {
...
+        } else {
+          receiver.type = subSchema.type;
+        }
+      }
+
+      if (subSchema.enum !== undefined) {
+        if (Array.isArray(subSchema.enum) && Array.isArray(receiver.enum)) {
+          receiver.enum = subSchema.enum.concat(...receiver.enum);
```
Comment (RomanHotsiy) [inline src/services/OpenAPIParser.ts:291]:
```suggestion
          receiver.enum = [...subSchema.enum, ...receiver.enum];
```

File: src/services/models/Schema.ts [hunk 4]
```diff
@@ -355,6 +361,44 @@ export class SchemaModel {
       return innerSchema;
     });
   }
+
+  private initConditionalOperators(schema: OpenAPISchema, parser: OpenAPIParser) {
+    const {
+      if: ifOperator,
+      else: elseOperator = {},
+      then: thenOperator = {},
+      ...clearSchema
```
Comment (RomanHotsiy) [inline src/services/models/Schema.ts:370]:
```suggestion
      ...restSchema
```

File: src/services/models/Schema.ts [hunk 5]
```diff
@@ -355,6 +361,44 @@ export class SchemaModel {
...
+    } = schema;
+    const ifOperatorTitle =
+      (ifOperator && ifOperator['x-displayName']) || ifOperator?.title || 'if';
+    const elseOperatorTitle =
+      (elseOperator && elseOperator['x-displayName']) || elseOperator?.title || 'else';
+    const groupedOperators = [
+      mergeObjects(
+        {},
+        clearSchema,
```
Comment (RomanHotsiy) [inline src/services/models/Schema.ts:379]:
```suggestion
        restSchema,
```

File: src/services/models/Schema.ts [hunk 6]
```diff
@@ -355,6 +361,44 @@ export class SchemaModel {
...
+        { allOf: [ifOperator, thenOperator] },
+        { title: ifOperatorTitle },
+      ),
+      mergeObjects({}, clearSchema, elseOperator, { title: elseOperatorTitle }),
+    ];
+    this.conditionalEnum = [ifOperatorTitle, elseOperatorTitle];
```
Comment (RomanHotsiy) [inline src/services/models/Schema.ts:385]: This looks redundant. Let's use oneOf only.

File: src/services/models/Schema.ts (hunk 5 above)
Comment (RomanHotsiy) [reply src/services/models/Schema.ts:379]:
```
{
  allOf: [clearSchema, ifOperator, thenOperator, { title: ifOperatorTitle }
}
```

```
{
  allOf: [clearSchema, elseSchema, { title: ifOperatorTitle }
}
```

File: src/services/models/Schema.ts [hunk 7]
```diff
@@ -355,6 +361,44 @@ export class SchemaModel {
...
+
+    this.oneOf = groupedOperators.map((variant, idx) => {
+      const merged = parser.mergeAllOf(parser.deref(variant || {}), this.pointer + '/oneOf/' + idx);
```
Comment (RomanHotsiy) [inline src/services/models/Schema.ts:388]: check if `parser.mergeAllOf` is needed, I think SchemaModel already merged allOf.

File: src/components/Fields/Field.tsx [hunk 8]
```diff
@@ -87,12 +91,24 @@ export class Field extends React.Component<FieldProps> {
...
+          field.schema.conditionalEnum && (
+            <ConditionalWrapper>
+              <DiscriminatorDropdown
+                parent={field.schema}
+                enumValues={field.schema.conditionalEnum}
+              />
+            </ConditionalWrapper>
+          )
+      : undefined;
+
```
Comment (RomanHotsiy) [inline src/components/Fields/Field.tsx:105]: Plz, remove.

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]

File: src/services/models/Schema.ts [hunk 9]
```diff
@@ -355,6 +361,44 @@ export class SchemaModel {
...
+      const title = merged.title || this.title;
+      const result = new SchemaModel(
+        parser,
+        {
+          ...merged,
```
Comment (RomanHotsiy) [inline src/services/models/Schema.ts:393]:
```suggestion
          ...variant,
```

File: src/utils/openapi.ts [hunk 10]
```diff
@@ -125,6 +125,10 @@ export function isPrimitiveType(
     return false;
   }
 
+  if (schema.if || schema.then || schema.else) {
```
Comment (RomanHotsiy) [inline src/utils/openapi.ts:128]: fix condition

**Review by RomanHotsiy — APPROVED** [review]

**Pull request merged**